### PR TITLE
STITCH-1210 - Add a withDocument overload to RequestOptions.encode

### PR DESCRIPTION
**Pull request by edaniels**
- Make some Swift 4.1 fixes

### CODE REVIEW

File: MongoDBService.podspec [hunk 1]
```diff
@@ -15,8 +15,8 @@ Pod::Spec.new do |s|
   				   }
   s.source_files  = "MongoDBService/MongoDBService/**/*.swift"
   s.requires_arc = true
-  s.dependency "PromiseKit/CorePromise", "~> 6.0.0"
+  s.dependency "PromiseKit/CorePromise", "~> 6.1.0"
   s.dependency "StitchLogger", "~> 2.0.0"
   s.dependency "ExtendedJson", "~> 2.0.0"
-  s.dependency "StitchCore", "~> 3.0.0"
+  s.dependency "StitchCore", "~> 3.0.1"
```
Comment (edaniels) [inline MongoDBService.podspec]: Was this right to do?

Comment (jsflax) [reply MongoDBService.podspec]: Yes. I'll complete the bump version script so it's clearer what needs to happen here.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 2]
```diff
@@ -348,7 +350,7 @@ public class StitchClient: StitchClientType {
     public func logout() -> Promise<Void> {
         if !isAuthenticated {
             printLog(.info, text: "Tried logging out while there was no authenticated user found.")
-            return Promise.init { $0.fulfill(()) }
+            return Promise.value(())
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [q] Do we need the extra parentheses? Looks a bit odd to the eye.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 3]
```diff
@@ -477,7 +479,7 @@ public class StitchClient: StitchClientType {
         return httpClient.doRequest {
             $0.method = .post
             $0.endpoint = self.routes.functionsCallRoute
-            try $0.encode(withData: ["name": name,
+            try $0.encode(withDocument: ["name": name,
                                      "arguments": BSONArray(array: args)])
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [style] Align keys

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 4]
```diff
@@ -495,7 +497,7 @@ public class StitchClient: StitchClientType {
         return httpClient.doRequest {
             $0.method = .post
             $0.endpoint = self.routes.functionsCallRoute
-            try $0.encode(withData: ["name": name,
+            try $0.encode(withDocument: ["name": name,
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [style] align keys

File: StitchCore/StitchCore/Source/Push/PushManager.swift [hunk 5]
```diff
@@ -6,7 +6,25 @@ import Foundation
...
+
+    private struct PushClientKey: Hashable {
+        let clientAppId: String
+        let providerName: PushProviderName
+
+        var hashValue: Int {
+            return "\(clientAppId) \(providerName)".hashValue
+        }
+
+        init(_ clientAppId: String, _ providerName: PushProviderName) {
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Push/PushManager.swift]: [minor] I'm not sure this is an appropriate use of unnamed parameters

File: StitchCore/StitchCoreTests/Admin/AuthProviders/AuthProvidersResources.swift [hunk 6]
```diff
@@ -24,7 +24,7 @@ internal struct AuthProviderResponse: Codable {
         let container = try decoder.container(keyedBy: CodingKeys.self)
 
         self.id = try container.decode(String.self, forKey: .id)
-        self.disabled = try container.decode(Int.self, forKey: .disabled) == 0 ? false : true
+        self.disabled = try container.decode(Bool.self, forKey: .disabled)
```
Comment (jsflax) [inline StitchCore/StitchCoreTests/Admin/AuthProviders/AuthProvidersResources.swift]: 👍

File: StitchCore/StitchCore/Source/Push/PushManager.swift [hunk 7]
```diff
@@ -18,22 +36,21 @@ public class PushManager: AuthDelegate {
...
+            return client
+        }
 
+        var pClient: PushClient! = nil
         switch info {
-        case let gcmInfo as StitchGCMPushProviderInfo: client =
-            StitchGCMPushClient(stitchClient: self.stitchClient,
+        case let gcmInfo as StitchGCMPushProviderInfo:
+            pClient = StitchGCMPushClient(stitchClient: self.stitchClient,
                                 info: gcmInfo)
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Push/PushManager.swift]: [style] align params

File: StitchCore/StitchCoreTests/Admin/StitchAdminClient.swift [hunk 8]
```diff
@@ -80,8 +80,12 @@ public class StitchAdminClient {
             $0.refreshOnFailure = false
             $0.useRefreshToken = false
         }.flatMap {
-            return try JSONDecoder().decode(UserProfile.self,
-                                            from: JSONSerialization.data(withJSONObject: $0))
+            do {
+                return try JSONDecoder().decode(UserProfile.self,
+                                         from: JSONSerialization.data(withJSONObject: $0))
```
Comment (jsflax) [inline StitchCore/StitchCoreTests/Admin/StitchAdminClient.swift]: [style] align params

File: StitchCore/StitchCore/Source/Core/StitchError.swift [hunk 9]
```diff
@@ -43,6 +43,7 @@ public enum StitchError: Error {
                  - errorCode: String based error code that should coincide with `ServerErrorReason`
                  - errorMessage: More precise error message
          */
+        // swiftlint:disable:next cyclomatic_complexity
         init(errorCode: String, errorMessage: String) {
             switch errorCode {
             case "InvalidSession":
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Core/StitchError.swift]: [nit][maybe later] Should probably pull the error codes out into an `enum`.

File: ExtendedJson/ExtendedJson/Source/Bson/Coders.swift [hunk 10]
```diff
@@ -638,7 +638,7 @@ fileprivate struct _BSONKeyedDecodingContainer<Key : CodingKey> : KeyedDecodingC
     var codingPath: [CodingKey]
 
     var allKeys: [Key] {
-        return decoder.target.document.orderedKeys.flatMap { Key(stringValue: $0 as! String) }
+        return decoder.target.document.orderedKeys.flatMap{ Key(stringValue: $0 as! String) }
```
[the remark this answers is not in the document]
Comment (edaniels) [reply ExtendedJson/ExtendedJson/Source/Bson/Coders.swift]: whoops. linter should've caught it

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 2 above)
Comment (edaniels) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: yes, you need some argument value, the value is a `Void`

File: StitchCore/StitchCore/Source/Push/PushManager.swift (hunk 7 above)
Comment (edaniels) [reply StitchCore/StitchCore/Source/Push/PushManager.swift]: wish it did this automatically

File: StitchCore/StitchCore/Source/Core/StitchError.swift (hunk 9 above)
Comment (edaniels) [reply StitchCore/StitchCore/Source/Core/StitchError.swift]: yeah probably. can't nit me for doing nothing :)

**Pull request merged**